lib/livebook/runtime/k8s.ex

```diff
@@ -0,0 +1,566 @@
+  # protocol (`-proto_dist`). We configure the Kubernetes node to use IPv6
+  # distribution (`-proto_dist inet6_tcp`). However, depending whether
+  # the local node runs IPv4 or IPv6 distribution, we configure the
+  # kubectl proxy to bind to a IPv4 or IPv6 loopback respectively. The
+  # proxy always communicates with the Kubernetes Pod over IPv6, as
+  # is the case with all internal networking. Consequently, regardless
+  # of the protocol used by the local node, the remote node perceives
+  # it as IPv6.
+  #
+  # Sidenote, a node using IPv6 distribution may accept connections
+  # from a node using IPv4, depending on the `:kernel` application
+  # configuration `inet_dist_listen_options` -> `ipv6_v6only`, which
+  # has OS-specific value. However, we don't rely on this here.
+
+  defstruct [:config, :node, :req, :server_pid, :lv_pid]
+
+  @type config :: %{
+          context: String.t(),
+          namespace: String.t(),
+          home_pvc: String.t(),
+          docker_tag: String.t(),
+          pod_template: String.t()
+        }
+
+  @type t :: %__MODULE__{
+          node: node() | nil,
+          req: Req.Request.t(),
+          server_pid: pid() | nil,
+          lv_pid: pid()
+        }
+
+  use GenServer, restart: :temporary
+
+  require Logger
+
+  @doc """
+  Returns a new runtime instance.
+  """
+  @spec new(config :: map(), req :: Req.Request.t()) :: t()
+  def new(config, req) do
+    %__MODULE__{config: config, req: req, lv_pid: self()}
+  end
+
+  def __connect__(runtime) do
+    {:ok, pid} =
+      DynamicSupervisor.start_child(Livebook.RuntimeSupervisor, {__MODULE__, {runtime, self()}})
+
+    pid
+  end
+
+  @doc false
+  def start_link({runtime, caller}) do
+    GenServer.start_link(__MODULE__, {runtime, caller})
+  end
+
+  @impl true
+  def init({runtime, caller}) do
+    state = %{primary_ref: nil}
+    {:ok, state, {:continue, {:init, runtime, caller}}}
+  end
+
+  @impl true
+  def handle_continue({:init, runtime, caller}, state) do
+    config = runtime.config
+    %{namespace: namespace, context: context} = config
+    req = runtime.req
+
+    kubeconfig =
+      if System.get_env("KUBERNETES_SERVICE_HOST"),
+        do: nil,
+        else: System.get_env("KUBECONFIG") || Path.join(System.user_home!(), ".kube/config")
+
+    cluster_data = get_cluster_data(kubeconfig)
+
+    runtime_data =
+      %{
+        node_base: cluster_data.node_base,
+        cookie: Node.get_cookie(),
+        dist_port: cluster_data.remote_port
+      }
+      |> :erlang.term_to_binary()
+      |> Base.encode64()
+
+    with {:ok, pod_name} <-
+           with_log(caller, "create pod", fn ->
+             create_pod(req, config, runtime_data, cluster_data.remote_port)
+           end),
+         _ <-
+           Phoenix.LiveView.send_update(
+             runtime.lv_pid,
+             LivebookWeb.SessionLive.K8sRuntimeComponent,
+             id: "runtime-config-k8s",
+             pod_name: pod_name
+           ),
```

Comment:
We should not communicate with the LV directly. I generally wouldn't use `send_update` outside LV code, because this introduces brittle coupling.

For a more practical reason, when the user refreshes the page this would no longer work (or for another user looking at the UI).

---

Let's remove this part. I need to think more how to handle cancellation from the UI. Technically, I think it would be fine to just kill the process, because the resources are designed to autoterminate (e.g. once the pod comes up it would quickly terminate itself).

Reply:
Yeah I've also noticed that when reconnecting using the button in the left sidebar, this doesn't really work. Left it in for the discussion. 

So you're proposing to kill the runtime process? Do we get its pid in the LV? And I guess I add a button to the sidebar as well?

Reply:
Let's skip this for now. I think it would also be useful for the Fly runtime, so I will have a look at that and then we can do the same here.

I think killing the process is the way to go, I just need to fit it into the runtime API. If we want to eagerly delete the pod on abort, then we can do still do that. We would spawn a watcher process that monitors the caller. Once the pod is created we send info to the watcher. When the caller terminates unexpectedly, the watcher can do eager cleanup.

Reply:
You want to skip the whole abort functionality for now? The problem is: It is quite easy to create a manifest for a pod that can't be scheduled by the Scheduler. Let's say I want to connect to my cluster but I'd like the node to have a GPU. So I add `nvidia.com/gpu: "1"` to my resources but often those nodes are tainted and I forgot to put the toleration for that taint. The pod cannot be scheduled and we have to wait the full timeout (5/10/15 minutes?). 

I can skip it in this PR. Just saying I think it should be addressed as this can lead to frustration.

Reply:
> You want to skip the whole abort functionality for now?

Yes, but my plan is to add abort for Fly runtime soon and then we mirror it in this PR or add afterwards, but definitely before the next release :)

Reply:
@mruoss see `fly.ex` and `fly_component.ex` in #2776, let's do the same here. If there are any issues, let me know!

Reply:
Nice! Will do.